### PR TITLE
Add support for parallel backup & restore

### PR DESCRIPTION
**Pull request by hossainemruz**

### CODE REVIEW

File: pkg/restic/config.go [hunk 1]
```diff
@@ -106,3 +106,30 @@ func (w *ResticWrapper) GetRepo() string {
 	}
 	return ""
 }
+
+// DeepCopy function copy input ResticWrapper and returns a new wrapper with copy of its content.
+func (in *ResticWrapper) DeepCopy() *ResticWrapper {
```
Comment (tamalsaha) [inline pkg/restic/config.go]: Why do we need this?

File: pkg/status/status.go [hunk 2]
```diff
@@ -36,10 +36,14 @@ func (o UpdateStatusOptions) UpdateBackupStatusFromFile() error {
 	if err != nil {
 		return err
 	}
+
 	var backupErr error
-	if backupOutput.HostBackupStats.Error != "" {
-		backupErr = fmt.Errorf(backupOutput.HostBackupStats.Error)
+	for _, hostStats := range backupOutput.HostBackupStats {
+		if hostStats.Error != "" {
+			backupErr = errors.NewAggregate([]error{backupErr, fmt.Errorf(hostStats.Error)})
```
Comment (tamalsaha) [inline pkg/status/status.go]: This is not right. We should collect all the errors in a `[]error`, then call 

`errors.NewAggregators(errs)`

File: pkg/status/status.go [hunk 3]
```diff
@@ -140,29 +153,35 @@ func (o UpdateStatusOptions) UpdatePostRestoreStatus(restoreOutput *restic.Resto
...
-	var eventType, eventReason, eventMessage string
-	if restoreOutput.HostRestoreStats.Error != "" {
-		eventType = core.EventTypeWarning
-		eventReason = eventer.EventReasonHostRestoreFailed
-		eventMessage = fmt.Sprintf("restore failed for host %q. Reason: %s", restoreOutput.HostRestoreStats.Hostname, restoreOutput.HostRestoreStats.Error)
-	} else {
-		eventType = core.EventTypeNormal
-		eventReason = eventer.EventReasonHostRestoreSucceeded
-		eventMessage = fmt.Sprintf("restore succeeded for host %q", restoreOutput.HostRestoreStats.Hostname)
+		_, err = stash_util_v1beta1.UpdateRestoreSessionStatusForHost(o.StashClient.StashV1beta1(), restoreSession, hostStats)
```
Comment (tamalsaha) [inline pkg/status/status.go]: Why are we not updating in one api call?

File: pkg/restic/output.go [hunk 4]
```diff
@@ -10,13 +10,12 @@ import (
 	"path/filepath"
 	"strings"
 
-	"github.com/appscode/go/types"
 	api_v1beta1 "stash.appscode.dev/stash/apis/stash/v1beta1"
 )
 
 type BackupOutput struct {
 	// HostBackupStats shows backup statistics of current host
```
Comment (tamalsaha) [inline pkg/restic/output.go]: `current host` ?

File: pkg/restic/metrics.go [hunk 5]
```diff
@@ -8,28 +8,30 @@ import (
...
 )
 
+// BackupMetrics defines prometheus metrics for backup setup and individual host backup
 type BackupMetrics struct {
 	// BackupSetupMetrics indicates whether backup was successfully setup for the target
 	BackupSetupMetrics prometheus.Gauge
-	// BackupSessionMetrics shows metrics related to last backup session
-	BackupSessionMetrics *BackupSessionMetrics
-	// RepositoryMetrics shows metrics related to repository after last backup
-	RepositoryMetrics *RepositoryMetrics
```
Comment (tamalsaha) [inline pkg/restic/metrics.go]: `RepositoryMetrics *RepositoryMetrics` where did this go?

File: pkg/restic/config.go [hunk 6]
```diff
@@ -106,3 +106,30 @@ func (w *ResticWrapper) GetRepo() string {
...
+
+		// set values in.sh to out.sh
+		for k, v := range in.sh.Env {
+			out.sh.Env[k] = v
+		}
+		// don't use same stdin, stdout, stderr for each instant to avoid data race.
+		//out.sh.Stdin = in.sh.Stdin
+		//out.sh.Stdout = in.sh.Stdout
+		//out.sh.Stderr = in.sh.Stderr
+		out.sh.ShowCMD = in.sh.ShowCMD
```
Comment (tamalsaha) [inline pkg/restic/config.go]: This is `shallow` copy.

File: pkg/restic/backup.go [hunk 7]
```diff
@@ -72,15 +59,197 @@ func (w *ResticWrapper) RunBackup(backupOption BackupOptions) (*BackupOutput, er
...
+	wg := sync.WaitGroup{}
+	// concurrencyLimiter channel is used to limit maximum number simultaneous go routine
+	concurrencyLimiter := make(chan bool, maxConcurrency)
+	defer close(concurrencyLimiter)
+
+	var backupErr error
+	mu := sync.Mutex{} // use lock to avoid racing condition
+	backupOutput := &BackupOutput{}
+
+	for _, opt := range backupOptions {
```
Comment (tamalsaha) [inline pkg/restic/backup.go]: The classic range pointer mistake. Use this:

`for i := range backupOptions`

File: pkg/restic/backup.go [hunk 8]
```diff
@@ -72,15 +59,197 @@ func (w *ResticWrapper) RunBackup(backupOption BackupOptions) (*BackupOutput, er
...
+				return
+			}
+			hostStats.Duration = time.Now().Sub(start).String()
+			hostStats.Phase = api_v1beta1.HostBackupSucceeded
+
+			// add hostStats to backupOutput. use lock to avoid racing condition.
+			mu.Lock()
+			backupOutput.upsertHostBackupStats(hostStats)
+			mu.Unlock()
+		}(opt,startTime)
```
Comment (tamalsaha) [inline pkg/restic/backup.go]: `} (backupOptions[i], time.Now())`

File: pkg/restic/backup.go [hunk 9]
```diff
@@ -72,15 +59,197 @@ func (w *ResticWrapper) RunBackup(backupOption BackupOptions) (*BackupOutput, er
...
+
+			// sh field in ResticWrapper is a pointer. we must not use same w in multiple go routine.
+			// otherwise they might enter in a racing condition.
+			nw := w.DeepCopy()
+
+			hostStats, err := nw.runBackup(opt)
+			if err != nil {
+				// acquire lock to make sure no other go routine is writing to backupErr variable
+				mu.Lock()
+				backupErr = errors.NewAggregate([]error{backupErr, err})
```
Comment (tamalsaha) [inline pkg/restic/backup.go]: `errors.NewAggregate` should be called outside. Here we should just append the errors.

File: pkg/restic/backup.go [hunk 10]
```diff
@@ -72,15 +59,197 @@ func (w *ResticWrapper) RunBackup(backupOption BackupOptions) (*BackupOutput, er
...
+				mu.Unlock()
+				return
+			}
+			hostStats.Duration = time.Now().Sub(start).String()
```
Comment (tamalsaha) [inline pkg/restic/backup.go]: Should we count time when the go routine was created or when it actually started running?

File: pkg/restic/config.go [hunk 11]
```diff
@@ -106,3 +106,30 @@ func (w *ResticWrapper) GetRepo() string {
...
+		out.sh.PipeFail = in.sh.PipeFail
+		out.sh.PipeStdErrors = in.sh.PipeStdErrors
+
+	}
+	out.config = in.config
```
Comment (tamalsaha) [inline pkg/restic/config.go]: This is `shallow` copy.

File: pkg/restic/restore.go [hunk 12]
```diff
@@ -1,57 +1,224 @@
...
+	}
+
+	err := w.runRestore(restoreOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Restore successful. Read current time and calculate total session duration.
+	endTime := time.Now()
+	restoreStats.Duration = endTime.Sub(startTime).String()
```
Comment (tamalsaha) [inline pkg/restic/restore.go]: time.Since(startTime)

File: pkg/restic/restore.go [hunk 13]
```diff
@@ -1,57 +1,224 @@
...
+	wg := sync.WaitGroup{}
+	// concurrencyLimiter channel is used to limit maximum number simultaneous go routine
+	concurrencyLimiter := make(chan bool, maxConcurrency)
+	defer close(concurrencyLimiter)
+
+	var restoreErr error
+	restoreOutput := &RestoreOutput{}
+	mu := sync.Mutex{}
+
+	for _, opt := range restoreOptions {
```
Comment (tamalsaha) [inline pkg/restic/restore.go]: `for i := range restoreOptions`

File: pkg/restic/restore.go [hunk 14]
```diff
@@ -1,57 +1,224 @@
...
+	wg := sync.WaitGroup{}
+	// concurrencyLimiter channel is used to limit maximum number simultaneous go routine
+	concurrencyLimiter := make(chan bool, maxConcurrency)
+	defer close(concurrencyLimiter)
+
+	var restoreErr error
+	restoreOutput := &RestoreOutput{}
+	mu := sync.Mutex{}
+
+	for _, opt := range dumpOptions {
```
Comment (tamalsaha) [inline pkg/restic/restore.go]: same

File: pkg/restic/backup.go (hunk 7 above)
Comment (hossainemruz) [reply pkg/restic/backup.go]: Not required. Here, we are sending opt as parameter so no pointer issue.

File: pkg/restic/restore.go [hunk 15]
```diff
@@ -1,57 +1,217 @@
...
+
+	// WaitGroup to wait until all go routine finish
+	wg := sync.WaitGroup{}
+	// concurrencyLimiter channel is used to limit maximum number simultaneous go routine
+	concurrencyLimiter := make(chan bool, maxConcurrency)
+	defer close(concurrencyLimiter)
+
+	var restoreErr []error
+	restoreOutput := &RestoreOutput{}
+	mu := sync.Mutex{}
```
Comment (tamalsaha) [inline pkg/restic/restore.go]: `var mu sync.Mutex`

**Pull request merged**